File: packages/core/src/modules/hassault/viewclips.ts
/**
 * What the hands do, as two layers of keyframed offsets.
 *
 * ## Offsets, not positions
 *
 * Every pose is an offset from the weapon's **grip anchor** (`arms.ts`,
 * `models/grips.json`), in the weapon's own space. That is what makes a partial
 * keyframe legal: a track that says nothing about a hand leaves it on the gun,
 * so `reload` can be written entirely in terms of the support hand and never
 * mention the trigger hand it is not moving.
 *
 * ## Two layers, and why they must be channel-disjoint
 *
 * The base layer is **locomotion** — idle, walk, run, jump, land — looping, and
 * phase-driven off the view model's own `bobPhase` so it cannot drift out of
 * step with the bob. The upper layer is **actions** — reload, inspect, throw,
 * draw — one-shot.
 *
 * The action layer **replaces** the channels it names; locomotion supplies the
 * rest. It is not a weighted average, and `models/clips.ts` already explains
 * why: two poses averaged on one bone give you half a reload, which is a motion
 * that belongs to neither animation. That file makes its track sets disjoint via
 * `isUpperBody`; this one does it per channel, which is the same idea with four
 * channels instead of a skeleton.
 *
 * ## `t` is a fraction, never seconds
 *
 * The reload track stretches to whatever `reloadTime` the server served, so one
 * authored motion serves a 1.4s pistol and a 2.6s shotgun and both come back up
 * on the frame the magazine is full. Written in seconds the two would need a
 * table nobody would keep in step with `weapons.py` — the same argument
 * `RELOAD_DIP_IN` makes.
 *
 * ## The seam
 *
 * `PoseSource` is the one place a different *kind* of animation could arrive.
 * The authored tracks below are one implementation; rigged clips off a skeleton
 * would be another, and nothing in `arms.ts` or the view model would change.
 * One interface, no speculative machinery.
 */
import clips from './models/viewclips.json';

/** The four channels a pose can move. Anything not named is left on the gun. */
export interface ArmPose {
  primary: [number, number, number];
  support: [number, number, number];
  primaryRoll: number;
  supportRoll: number;
}

export type PartialPose = Partial<ArmPose>;

export interface Keyframe {
  /** Normalised over the clip's own length, 0..1. Never seconds. */
  t: number;
  pose: PartialPose;
}

export type Track = Keyframe[];

export type LocomotionClip = 'idle' | 'walk' | 'run' | 'jump' | 'land';
export type ActionClip = 'reload' | 'inspect' | 'throw' | 'draw';

interface ClipFile {
  locomotion: Record<string, Track>;
  actions: Record<string, Track>;
}

const CLIPS = clips as unknown as ClipFile;

/** How long the land dip lasts, in seconds. Short — it is a jolt, not a pose. */
export const LAND_DURATION = 0.22;
/** How long a draw takes. Matches the view model's own `DRAW_TIME`. */
export const DRAW_DURATION = 0.25;
/** The grenade throw's three phases, end to end. */
export const THROW_DURATION = 0.55;

/**
 * Which locomotion clip this frame wants.
 *
 * Airborne beats everything: a player is either on the ground moving or they are
 * not, and a walk cycle in mid-air reads as a bug — the same call
 * `viewmodel.update` already makes about the bob.
 */
export function selectLocomotion(state: {
  speed: number;
  moveSpeed: number;
  onGround: boolean;
  sinceLanded: number;
}): LocomotionClip {
  if (!state.onGround) return 'jump';
  if (state.sinceLanded < LAND_DURATION) return 'land';
  const fraction = state.moveSpeed > 0 ? state.speed / state.moveSpeed : 0;
  if (fraction < 0.15) return 'idle';
  return fraction < 0.7 ? 'walk' : 'run';
}

/** How long a crossfade into this clip should take, in seconds. */
export function fadeFor(clip: LocomotionClip | ActionClip): number {
  // A landing is an impact and must not be eased into — it is the one clip whose
  // whole content is the suddenness.
  if (clip === 'land') return 0;
  if (clip === 'jump') return 0.08;
  return 0.14;
}

function lerp3(
  a: [number, number, number],
  b: [number, number, number],
  t: number,
): [number, number, number] {
  return [a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t, a[2] + (b[2] - a[2]) * t];
}

const ZERO3: [number, number, number] = [0, 0, 0];

/**
 * Sample one track at a normalised time.
 *
 * A channel absent from **both** surrounding keyframes stays absent from the
 * result — that is the whole of "a partial track leaves that hand alone". A
 * channel present in only one is interpolated against zero, since zero is the
 * grip anchor itself.
 */
export function sampleTrack(track: Track, t: number): PartialPose {
  if (track.length === 0) return {};
  const clamped = Math.max(0, Math.min(1, t));
  if (track.length === 1 || clamped <= track[0].t) return { ...track[0].pose };
  const last = track[track.length - 1];
  if (clamped >= last.t) return { ...last.pose };

  let i = 0;
  while (i < track.length - 2 && track[i + 1].t < clamped) i += 1;
  const a = track[i];
  const b = track[i + 1];
  const span = b.t - a.t;
  const k = span <= 1e-9 ? 0 : (clamped - a.t) / span;

  const out: PartialPose = {};
  if (a.pose.primary !== undefined || b.pose.primary !== undefined) {
    out.primary = lerp3(a.pose.primary ?? ZERO3, b.pose.primary ?? ZERO3, k);
  }
  if (a.pose.support !== undefined || b.pose.support !== undefined) {
    out.support = lerp3(a.pose.support ?? ZERO3, b.pose.support ?? ZERO3, k);
  }
  if (a.pose.primaryRoll !== undefined || b.pose.primaryRoll !== undefined) {
    const from = a.pose.primaryRoll ?? 0;
    out.primaryRoll = from + ((b.pose.primaryRoll ?? 0) - from) * k;
  }
  if (a.pose.supportRoll !== undefined || b.pose.supportRoll !== undefined) {
    const from = a.pose.supportRoll ?? 0;
    out.supportRoll = from + ((b.pose.supportRoll ?? 0) - from) * k;
  }
  return out;
}

/**
 * Lay an action over a locomotion pose.
 *
 * **Replace, not blend.** A channel the action names is the action's; everything
 * else is locomotion's. See the module header for why averaging them is wrong.
 */
export function mergePose(base: PartialPose, action: PartialPose): PartialPose {
  return { ...base, ...action };
}

/** Where a pose source's poses come from. See the module header's seam note. */
export interface PoseSource {
  locomotion(clip: LocomotionClip, phase: number): PartialPose;
  action(clip: ActionClip, t: number): PartialPose;
}

/** The keyframed tracks in `models/viewclips.json`. The only source today. */
export class AuthoredPoseSource implements PoseSource {
  locomotion(clip: LocomotionClip, phase: number): PartialPose {
    const track = CLIPS.locomotion[clip];
    if (!track) return {};
    // Wrapped rather than clamped: locomotion loops, and a phase that ran past 1
    // would freeze the cycle at its last keyframe.
    return sampleTrack(track, phase - Math.floor(phase));
  }

  action(clip: ActionClip, t: number): PartialPose {
    const track = CLIPS.actions[clip];
    return track ? sampleTrack(track, t) : {};
  }
}

/** Every clip name the file actually carries. Exported for the tests. */
export const LOCOMOTION_CLIPS = Object.keys(CLIPS.locomotion).filter(
  (k) => !k.startsWith('_'),
) as LocomotionClip[];
export const ACTION_CLIPS = Object.keys(CLIPS.actions).filter(
  (k) => !k.startsWith('_'),
) as ActionClip[];

File: apps/native-fps/src/viewclips.rs
//! What the hands do, as two layers of keyframed offsets.
//!
//! The browser's `viewclips.ts`, reading the **same** `models/viewclips.json` —
//! compiled in with `include_str!` rather than retyped, because two hand-typed
//! copies of forty keyframes drift on the very first tweak.
//!
//! ## Offsets, not positions
//!
//! Every pose is an offset from the weapon's grip anchor (`arms.rs`), in the
//! weapon's own space. That is what makes a partial keyframe legal: a track that
//! says nothing about a hand leaves it on the gun, so `reload` is written
//! entirely in terms of the support hand and never mentions the trigger hand it
//! is not moving.
//!
//! ## Two layers, and why they must be channel-disjoint
//!
//! Locomotion is the base layer — looping, phase-driven off the view model's own
//! bob so it cannot drift out of step with it. Actions are the upper layer, and
//! they **replace** the channels they name rather than blending. `clips.rs`
//! already makes that argument for the third-person rig: two poses averaged on
//! one bone give you half a reload, a motion belonging to neither animation.
//!
//! ## `t` is a fraction, never seconds
//!
//! The reload track stretches to whatever `reloadTime` the server served — the
//! same argument `RELOAD_DIP_IN` makes, and the reason one authored motion
//! serves a 1.4s pistol and a 2.6s shotgun.

use std::sync::OnceLock;

use glam::Vec3;

const CLIPS_JSON: &str =
    include_str!("../../../packages/core/src/modules/hassault/models/viewclips.json");

/// One frame of hand motion. Every channel is optional: absent means "leave that
/// hand where the grip anchor put it".
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct ArmPose {
    pub primary: Option<Vec3>,
    pub support: Option<Vec3>,
    pub primary_roll: Option<f32>,
    pub support_roll: Option<f32>,
}

impl ArmPose {
    /// Lay an action over a locomotion pose.
    ///
    /// **Replace, not blend** — see the module header.
    pub fn merged_with(self, action: ArmPose) -> ArmPose {
        ArmPose {
            primary: action.primary.or(self.primary),
            support: action.support.or(self.support),
            primary_roll: action.primary_roll.or(self.primary_roll),
            support_roll: action.support_roll.or(self.support_roll),
        }
    }

    /// Scale every channel. Used to fade a walk cycle in with the bob.
    pub fn scaled(self, k: f32) -> ArmPose {
        ArmPose {
            primary: self.primary.map(|v| v * k),
            support: self.support.map(|v| v * k),
            primary_roll: self.primary_roll.map(|r| r * k),
            support_roll: self.support_roll.map(|r| r * k),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Locomotion {
    Idle,
    Walk,
    Run,
    Jump,
    Land,
}

impl Locomotion {
    fn key(self) -> &'static str {
        match self {
            Locomotion::Idle => "idle",
            Locomotion::Walk => "walk",
            Locomotion::Run => "run",
            Locomotion::Jump => "jump",
            Locomotion::Land => "land",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Action {
    Reload,
    Inspect,
    Throw,
    Draw,
}

impl Action {
    fn key(self) -> &'static str {
        match self {
            Action::Reload => "reload",
            Action::Inspect => "inspect",
            Action::Throw => "throw",
            Action::Draw => "draw",
        }
    }
}

/// How long the land dip lasts, in seconds. Short — it is a jolt, not a pose.
pub const LAND_DURATION: f32 = 0.22;
/// How long a draw takes. Matches the view model's own `DRAW_TIME`.
pub const DRAW_DURATION: f32 = 0.25;
/// The grenade throw's three phases, end to end.
pub const THROW_DURATION: f32 = 0.55;

#[derive(Debug, Clone, Copy)]
struct Keyframe {
    t: f32,
    pose: ArmPose,
}

type Track = Vec<Keyframe>;

struct Clips {
    locomotion: std::collections::HashMap<String, Track>,
    actions: std::collections::HashMap<String, Track>,
}

fn parse_pose(value: &serde_json::Value) -> ArmPose {
    let vec3 = |key: &str| -> Option<Vec3> {
        let a = value.get(key)?.as_array()?;
        Some(Vec3::new(
            a.first()?.as_f64()? as f32,
            a.get(1)?.as_f64()? as f32,
            a.get(2)?.as_f64()? as f32,
        ))
    };
    ArmPose {
        primary: vec3("primary"),
        support: vec3("support"),
        primary_roll: value
            .get("primaryRoll")
            .and_then(|v| v.as_f64())
            .map(|v| v as f32),
        support_roll: value
            .get("supportRoll")
            .and_then(|v| v.as_f64())
            .map(|v| v as f32),
    }
}

fn parse_tracks(value: &serde_json::Value) -> std::collections::HashMap<String, Track> {
    let mut out = std::collections::HashMap::new();
    let Some(map) = value.as_object() else {
        return out;
    };
    for (name, track) in map {
        // The file documents itself in `_comment` keys, which are not clips.
        if name.starts_with('_') {
            continue;
        }
        let Some(frames) = track.as_array() else {
            continue;
        };
        out.insert(
            name.clone(),
            frames
                .iter()
                .map(|f| Keyframe {
                    t: f["t"].as_f64().unwrap_or(0.0) as f32,
                    pose: parse_pose(&f["pose"]),
                })
                .collect(),
        );
    }
    out
}

fn clips() -> &'static Clips {
    static CLIPS: OnceLock<Clips> = OnceLock::new();
    CLIPS.get_or_init(|| {
        let file: serde_json::Value = serde_json::from_str(CLIPS_JSON).unwrap_or_default();
        Clips {
            locomotion: parse_tracks(&file["locomotion"]),
            actions: parse_tracks(&file["actions"]),
        }
    })
}

fn lerp_opt(a: Option<Vec3>, b: Option<Vec3>, k: f32) -> Option<Vec3> {
    match (a, b) {
        (None, None) => None,
        // A channel present in only one keyframe interpolates against zero,
        // since zero *is* the grip anchor.
        (from, to) => {
            let from = from.unwrap_or(Vec3::ZERO);
            Some(from + (to.unwrap_or(Vec3::ZERO) - from) * k)
        }
    }
}

fn lerp_opt_f32(a: Option<f32>, b: Option<f32>, k: f32) -> Option<f32> {
    match (a, b) {
        (None, None) => None,
        (from, to) => {
            let from = from.unwrap_or(0.0);
            Some(from + (to.unwrap_or(0.0) - from) * k)
        }
    }
}

/// Sample one track at a normalised time.
///
/// A channel absent from **both** surrounding keyframes stays absent — that is
/// the whole of "a partial track leaves that hand alone".
fn sample(track: &Track, t: f32) -> ArmPose {
    if track.is_empty() {
        return ArmPose::default();
    }
    let t = t.clamp(0.0, 1.0);
    if track.len() == 1 || t <= track[0].t {
        return track[0].pose;
    }
    let last = track[track.len() - 1];
    if t >= last.t {
        return last.pose;
    }
    let mut i = 0;
    while i < track.len() - 2 && track[i + 1].t < t {
        i += 1;
    }
    let (a, b) = (track[i], track[i + 1]);
    let span = b.t - a.t;
    let k = if span <= 1e-9 { 0.0 } else { (t - a.t) / span };
    ArmPose {
        primary: lerp_opt(a.pose.primary, b.pose.primary, k),
        support: lerp_opt(a.pose.support, b.pose.support, k),
        primary_roll: lerp_opt_f32(a.pose.primary_roll, b.pose.primary_roll, k),
        support_roll: lerp_opt_f32(a.pose.support_roll, b.pose.support_roll, k),
    }
}

/// The looping base layer, at a phase. Wrapped rather than clamped: a phase that
/// ran past 1 would freeze the cycle at its last keyframe.
pub fn locomotion_pose(clip: Locomotion, phase: f32) -> ArmPose {
    match clips().locomotion.get(clip.key()) {
        Some(track) => sample(track, phase - phase.floor()),
        None => ArmPose::default(),
    }
}

/// The one-shot upper layer, at a fraction of its own length.
pub fn action_pose(clip: Action, t: f32) -> ArmPose {
    match clips().actions.get(clip.key()) {
        Some(track) => sample(track, t),
        None => ArmPose::default(),
    }
}

/// Which locomotion clip this frame wants.
///
/// Airborne beats everything: a walk cycle in mid-air reads as a bug — the same
/// call the view model already makes about the bob.
pub fn select_locomotion(
    speed: f32,
    move_speed: f32,
    on_ground: bool,
    since_landed: f32,
) -> Locomotion {
    if !on_ground {
        return Locomotion::Jump;
    }
    if since_landed < LAND_DURATION {
        return Locomotion::Land;
    }
    let fraction = if move_speed > 0.0 {
        speed / move_speed
    } else {
        0.0
    };
    if fraction < 0.15 {
        Locomotion::Idle
    } else if fraction < 0.7 {
        Locomotion::Walk
    } else {
        Locomotion::Run
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const LOCOMOTIONS: [Locomotion; 5] = [
        Locomotion::Idle,
        Locomotion::Walk,
        Locomotion::Run,
        Locomotion::Jump,
        Locomotion::Land,
    ];
    const ACTIONS: [Action; 4] = [Action::Reload, Action::Inspect, Action::Throw, Action::Draw];

    #[test]
    fn every_clip_the_file_promises_is_actually_there() {
        // A missing track samples to "no offset", which is an animation that
        // silently does nothing rather than an error.
        for clip in LOCOMOTIONS {
            assert!(
                clips().locomotion.contains_key(clip.key()),
                "no {} track",
                clip.key()
            );
        }
        for clip in ACTIONS {
            assert!(
                clips().actions.contains_key(clip.key()),
                "no {} track",
                clip.key()
            );
        }
    }

    #[test]
    fn every_locomotion_clip_loops_continuously() {
        // `t = 0` and `t = 1` must be the same pose, or the cycle ticks once per
        // stride — visible, but easy to mistake for the bob.
        for clip in LOCOMOTIONS {
            assert_eq!(
                locomotion_pose(clip, 0.0),
                locomotion_pose(clip, 1.0),
                "{} does not loop",
                clip.key()
            );
        }
    }

    #[test]
    fn every_action_ends_at_rest() {
        // An action that ends somewhere other than the grip leaves the hand
        // there until the next one runs.
        for clip in ACTIONS {
            assert_eq!(
                action_pose(clip, 1.0),
                ArmPose::default(),
                "{} does not end at rest",
                clip.key()
            );
        }
    }

    #[test]
    fn an_action_replaces_only_the_channels_it_names() {
        let walking = ArmPose {
            primary: Some(Vec3::ONE),
            support: Some(Vec3::splat(2.0)),
            ..ArmPose::default()
        };
        let reloading = ArmPose {
            support: Some(Vec3::splat(9.0)),
            ..ArmPose::default()
        };
        let merged = walking.merged_with(reloading);
        assert_eq!(merged.primary, Some(Vec3::ONE));
        assert_eq!(merged.support, Some(Vec3::splat(9.0)));
    }

    #[test]
    fn airborne_beats_everything() {
        assert_eq!(select_locomotion(20.0, 20.0, false, 99.0), Locomotion::Jump);
    }

    #[test]
    fn the_landing_dip_plays_and_then_stops() {
        assert_eq!(select_locomotion(0.0, 20.0, true, 0.0), Locomotion::Land);
        assert_eq!(
            select_locomotion(0.0, 20.0, true, LAND_DURATION + 0.01),
            Locomotion::Idle
        );
    }

    #[test]
    fn speed_picks_the_gait() {
        assert_eq!(select_locomotion(0.0, 20.0, true, 99.0), Locomotion::Idle);
        assert_eq!(select_locomotion(8.0, 20.0, true, 99.0), Locomotion::Walk);
        assert_eq!(select_locomotion(20.0, 20.0, true, 99.0), Locomotion::Run);
    }
}
